src/encoded/upgrade/genetic_modification.py

```diff
@@ -185,4 +185,78 @@ def genetic_modification_5_6(value, system):
 def genetic_modification_6_7(value, system):
     # https://encodedcc.atlassian.net/browse/ENCD-4042
     if value['purpose'] == 'validation':
-        value['purpose'] = 'characterization'
\ No newline at end of file
+        value['purpose'] = 'characterization'
+
+
+@upgrade_step('genetic_modification', '7', '8')
+def genetic_modification_7_8(value, system):
+    # https://encodedcc.atlassian.net/browse/ENCD-4349
+    import re
+    import copy
+    import time
+    import json
+
+    conn = system['registry'][CONNECTION]
+
+    sources = {
+        'addgene': r'^\d{5,6}$',
```

Comment:
Did you look at re.pattern?  I think you can create all these regex experssion in a pattern the then use them directly....
```
if sources[source_from_reagent].match(regex, identifier) is not None:
   etc...
```

Reply:
In this case, I want to preserve the mapping between source name and the appropriate regular expression. It could be done with pattern, but then I would have to name the groups in the regex to preserve the information (?P<id>), then extract the group names indirectly via Pattern.groupindex. This seems like a more straightforward way to do keep the source information